File: src/main/scala/com/xebia/functional/chains/retrievalqa/VectorQAChain.scala
package com.xebia.functional.chains.retrievalqa

import cats.effect.Sync
import cats.syntax.all.*

import com.xebia.functional.chains.combine.CombineDocumentsChain
import com.xebia.functional.chains.combine.StuffChain
import com.xebia.functional.chains.models.CombineDocumentsChainType
import com.xebia.functional.chains.models.CombineDocumentsChainType.Stuff
import com.xebia.functional.chains.models.Config
import com.xebia.functional.chains.models.InvalidChainInputsError
import com.xebia.functional.domain.Document
import com.xebia.functional.llm.openai.OpenAIClient
import com.xebia.functional.prompt.PromptTemplate
import com.xebia.functional.vectorstores.VectorStore
import eu.timepit.refined.types.string.NonEmptyString

class VectorQAChain[F[_]: Sync](
    llm: OpenAIClient[F],
    vectorStore: VectorStore[F],
    chainType: String,
    numberOfDocs: Int,
    llmModel: String,
    user: String,
    echo: Boolean,
    n: Int,
    temperature: Double,
    outputVariable: NonEmptyString,
    maxTokens: Int,
    onlyOutput: Boolean
) extends RetrievalQAChain[F]:
  val documentVariableName: String = "context"
  val inputVariable: String = "question"
  val config = Config(Set(inputVariable), Set(outputVariable.value), onlyOutput)
  val promptTemplate: F[PromptTemplate[F]] = QAPrompt.promptTemplate[F]

  def loadCombineDocsChain(
      chain: CombineDocumentsChainType,
      documents: List[Document],
      prompt: PromptTemplate[F]
  ): CombineDocumentsChain[F] =
    chain match
      case Stuff =>
        StuffChain.make[F](documents, llm, prompt, documentVariableName, llmModel, user, echo, n, temperature, outputVariable, maxTokens, onlyOutput)

  def getDocs(question: String): F[List[Document]] =
    vectorStore.similaritySearch(question, numberOfDocs)

  def call(inputs: Map[String, String]): F[Map[String, String]] =
    for
      cct <- CombineDocumentsChainType.fromString[F](chainType)
      prt <- promptTemplate
      question <- inputs.get(inputVariable).liftTo[F](InvalidChainInputsError(config.inputKeys, inputs))
      documents <- getDocs(question)
      combineChain = loadCombineDocsChain(cct, documents, prt)
      output <- combineChain.run(inputs)
    yield output

object VectorQAChain:
  def make[F[_]: Sync](
      llm: OpenAIClient[F],
      vectorStore: VectorStore[F],
      chainType: String,
      numberOfDocs: Int,
      llmModel: String,
      user: String,
      echo: Boolean,
      n: Int,
      temperature: Double,
      outputVariable: NonEmptyString,
      maxTokens: Int,
      onlyOutput: Boolean
  ): VectorQAChain[F] =
    new VectorQAChain[F](llm, vectorStore, chainType, numberOfDocs, llmModel, user, echo, n, temperature, outputVariable, maxTokens, onlyOutput)

  def makeWithDefaults[F[_]: Sync](
      llm: OpenAIClient[F],
      vectorStore: VectorStore[F],
      user: String,
      outputVariable: NonEmptyString
  ): VectorQAChain[F] =
    new VectorQAChain[F](
      llm,
      vectorStore,
      chainType = "stuff",
      numberOfDocs = 10,
      llmModel = "text-davinci-003",
      user = user,
      echo = false,
      n = 1,
      temperature = 0.0,
      outputVariable = outputVariable,
      maxTokens = 100,
      onlyOutput = true
    )

File: core/src/commonMain/kotlin/com/xebia/functional/chains/VectorQAChain.kt
package com.xebia.functional.chains

import arrow.core.Either
import arrow.core.raise.Raise
import arrow.core.raise.either
import arrow.core.raise.recover
import com.xebia.functional.AIError
import com.xebia.functional.AIError.Chain.InvalidTemplate
import com.xebia.functional.llm.openai.LLMModel
import com.xebia.functional.llm.openai.OpenAIClient
import com.xebia.functional.prompt.PromptTemplate
import com.xebia.functional.vectorstores.VectorStore

interface VectorQAChain : Chain {
  suspend fun getDocs(question: String): List<String>
}

@Suppress("LongParameterList")
suspend fun VectorQAChain(
  llm: OpenAIClient,
  llmModel: LLMModel,
  vectorStore: VectorStore,
  numOfDocs: Int,
  outputVariable: String,
  chainOutput: Chain.ChainOutput = Chain.ChainOutput.OnlyOutput
): VectorQAChain =
  object : VectorQAChain {

    private val documentVariableName: String = "context"
    private val inputVariable: String = "question"

    override val config: Chain.Config =
      Chain.Config(setOf(inputVariable), setOf(outputVariable), chainOutput)

    override suspend fun getDocs(question: String): List<String> =
      vectorStore.similaritySearch(question, numOfDocs)

    override suspend fun call(
      inputs: Map<String, String>
    ): Either<AIError.Chain, Map<String, String>> = either {
      val promptTemplate = promptTemplate()

      val question = validateInput(inputs, inputVariable)
      val documents = getDocs(question)

      val chain =
        CombineDocsChain(
          llm,
          promptTemplate,
          llmModel,
          documents,
          documentVariableName,
          outputVariable,
          chainOutput
        )

      chain.run(inputs).bind()
    }

    private fun Raise<InvalidTemplate>.promptTemplate(): PromptTemplate<String> =
      recover({
        val template =
          """
                |Use the following pieces of context to answer the question at the end. If you don't know the answer, just say that you don't know, don't try to make up an answer.
                |
                |{context}
                |
                |Question: {question}
                |Helpful Answer:"""
            .trimMargin()

        PromptTemplate(template, listOf("context", "question"))
      }) {
        raise(InvalidTemplate(it.reason))
      }
  }
